File: csharp/BuildPipelineTests/PipelineApprovalTest.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Xunit;
using VerifyTests;
using VerifyXunit;
using UntangledConditionals;

namespace BuildPipelineTests;

public class PipelineApprovalTest
{
    [Fact]
    public async Task PassingTests_EmailEnabled_DeploymentSuccessful()
    {
        var output = DoPipelineRun(TestStatus.PassingTests, true, true);
        await Verifier.Verify(output);
    }

    [Fact]
    public async Task PassingTests_EmailEnabled_DeploymentFailed()
    {
        var output = DoPipelineRun(TestStatus.PassingTests, true, false);
        await Verifier.Verify(output);
    }

    [Fact]
    public async Task PassingTests_EmailDisabled_DeploymentSuccessful()
    {
        var output = DoPipelineRun(TestStatus.PassingTests, false, true);
        await Verifier.Verify(output);
    }

    [Fact]
    public async Task NoTests_EmailEnabled_DeploymentSuccessful()
    {
        var output = DoPipelineRun(TestStatus.NoTests, true, true);
        await Verifier.Verify(output);
    }

    [Fact]
    public async Task NoTests_EmailEnabled_DeploymentFailed()
    {
        var output = DoPipelineRun(TestStatus.NoTests, true, false);
        await Verifier.Verify(output);
    }

    [Fact]
    public async Task FailingTests_EmailEnabled_DeploymentSuccessful()
    {
        var output = DoPipelineRun(TestStatus.FailingTests, true, true);
        await Verifier.Verify(output);
    }

    private string DoPipelineRun(TestStatus testStatus, bool sendSummary, bool buildsSuccessfully)
    {
        var spy = new StringBuilder();
        var config = new DefaultConfig(sendSummary);
        var emailer = new SpyEmailer(spy);
        var log = new SpyLogger(spy);
        var pipeline = new Pipeline(config, emailer, log);

        var project = Project.Builder()
            .SetTestStatus(testStatus)
            .SetDeploysSuccessfully(buildsSuccessfully)
            .Build();

        pipeline.Run(project);
        return spy.ToString();
    }

    private class DefaultConfig : Config
    {
        private readonly bool _sendEmail;
        public DefaultConfig(bool sendEmail) => _sendEmail = sendEmail;
        public bool SendEmailSummary() => _sendEmail;
    }

    private class SpyEmailer : Emailer
    {
        private readonly StringBuilder _spy;
        public SpyEmailer(StringBuilder spy) => _spy = spy;
        public void Send(string message)
        {
            _spy.AppendLine($"Email message: {message}");
        }
    }

    private class SpyLogger : Logger
    {
        private readonly StringBuilder _spy;
        public SpyLogger(StringBuilder spy) => _spy = spy;

        public void Info(string message)
        {
            _spy.AppendLine($"INFO: {message}");
        }

        public void Error(string message)
        {
            _spy.AppendLine($"ERROR: {message}");
        }
    }
}

File: cpp/test-catch2/PipelineApprovalTest.cpp
#include "ApprovalTests.hpp"
#include "catch2/catch.hpp"
#include "Pipeline.h"
#include <sstream>

using namespace std;

namespace {
    class SpyEmailer : public Emailer {
        stringstream &spy;
    public:
        explicit SpyEmailer(stringstream &spy) : spy(spy) {}
        void send(const string &message) override {
            spy << "Email message: " << message << "\n";
        }
    };

    class SpyLogger : public Logger {
        stringstream &spy;
    public:
        explicit SpyLogger(stringstream &spy) : spy(spy) {}
        void info(const string &message) override {
            spy << "INFO: " << message << "\n";
        }
        void error(const string &message) override {
            spy << "ERROR: " << message << "\n";
        }
    };

    string doPipelineRun(TestStatus testStatus, bool sendSummary, bool buildsSuccessfully) {
        stringstream spy;
        Config config(sendSummary);
        SpyEmailer emailer(spy);
        SpyLogger log(spy);
        Pipeline pipeline(config, emailer, log);

        Project project(buildsSuccessfully, testStatus);
        pipeline.run(project);

        return spy.str();
    }
}

TEST_CASE("PipelineRun") {
    SECTION("PassingTests_EmailEnabled_DeploymentSuccessful") {
        auto output = doPipelineRun(PASSING_TESTS, true, true);
        ApprovalTests::Approvals::verify(output);
    }

    SECTION("PassingTests_EmailEnabled_DeploymentFailed") {
        auto output = doPipelineRun(PASSING_TESTS, true, false);
        ApprovalTests::Approvals::verify(output);
    }

    SECTION("PassingTests_EmailDisabled_DeploymentSuccessful") {
        auto output = doPipelineRun(PASSING_TESTS, false, true);
        ApprovalTests::Approvals::verify(output);
    }

    SECTION("NoTests_EmailEnabled_DeploymentSuccessful") {
        auto output = doPipelineRun(NO_TESTS, true, true);
        ApprovalTests::Approvals::verify(output);
    }

    SECTION("NoTests_EmailEnabled_DeploymentFailed") {
        auto output = doPipelineRun(NO_TESTS, true, false);
        ApprovalTests::Approvals::verify(output);
    }

    SECTION("FailingTests_EmailEnabled_DeploymentSuccessful") {
        auto output = doPipelineRun(FAILING_TESTS, true, true);
        ApprovalTests::Approvals::verify(output);
    }
}
